feat(cip2): Input Selection

mkazlauskas opened the pull request:
# Context

[CIP2](https://cips.cardano.org/cips/cip2/)

# Proposed Solution

Implement Round-Robin Random Improve.

# Important Changes Introduced

- Create new package: `@cardano-sdk/cip2`
- Move BigIntMath from `@cardano-sdk/golden-test-generator` to `@cardano-sdk/core`
- Add some util functions to `cardano-sdk/core`
- Move OgmiosToCardanoWasm to `cardano-sdk/core`

rhyslbw commented on packages/cip2/test/RandomImprove.test.ts:
```diff
@@ -0,0 +1,199 @@
+  let utils: TestUtils;
+  beforeAll(() =>
+    // eslint-disable-next-line promise/always-return
+    loadCardanoSerializationLib().then((lib) => {
```
We tend to prefer async/await syntax in most places

rhyslbw replied:
We tend to prefer async/await syntax in most places

rhyslbw commented on packages/cip2/README.md:
```diff
@@ -0,0 +1,10 @@
+TODO: rename coin selection to input selection
+
+[cip-0002]: https://cips.cardano.org/cips/cip2/
+[borrowed]: https://github.com/Berry-Pool/nami-wallet/blob/8d3d2bdfd81b4a16b19df3da978ba578164b663c/src/lib/coinSelection.js
```
```suggestion
[referenced]: https://github.com/Berry-Pool/nami-wallet/blob/8d3d2bdfd81b4a16b19df3da978ba578164b663c/src/lib/coinSelection.js
```

rhyslbw commented on packages/cip2/package.json:
```diff
@@ -0,0 +1,28 @@
+    "test:debug": "DEBUG=true yarn test"
+  },
+  "devDependencies": {
+		"@emurgo/cardano-serialization-lib-nodejs": "^8.0.0",
```
```suggestion
    "@emurgo/cardano-serialization-lib-nodejs": "^8.0.0",
```

rhyslbw commented on packages/cip2/test/RandomImprove.test.ts:
```diff
@@ -0,0 +1,199 @@
+  minUtxoValue: 1_000_000
+};
+
+// TODO: Conservation of Outputs and Conservation of UTxO tests are duplicated exactly.
```
Yes this is a good refactor

rhyslbw commented on packages/cip2/src/RandomImprove.ts:
```diff
@@ -177,7 +29,8 @@ interface ImproveRange {
 
 /**
- * RandomImprove algorithm implementation for Input Selection
+ * RandomImprove algorithm implementation for Input Selection.
```
:+1:

rhyslbw commented on packages/cip2/src/RandomImprove.ts:
```diff
@@ -177,7 +29,8 @@ interface ImproveRange {
 /**
- * RandomImprove algorithm implementation for Input Selection
+ * RandomImprove algorithm implementation for Input Selection.
+ * See more details [here](https://cips.cardano.org/cips/cip2/#randomimprove).
```
:+1:

rhyslbw reviewed (changes requested):
Please drop the commits cherry-picked in #87, then rebase on `master`

mkazlauskas commented on packages/cip2/src/tsconfig.json:
```diff
@@ -0,0 +1,7 @@
+  "compilerOptions": {
+    "outDir": "../dist/",
+    "rootDir": "."
+  }
```
Could you please explain the purpose of this? Is it to automatically rebuild `cardano-serialization-lib` package?

rhyslbw replied:
Yes it's a [project reference](https://www.typescriptlang.org/docs/handbook/project-references.html)

rhyslbw commented on packages/cip2/src/types.ts:
```diff
@@ -0,0 +1,73 @@
+     * From the point of view of a wallet, this represents the value
+     * that has been selected from the wallet in order to cover the total payment value.
+     */
+    inputs: TransactionUnspentOutput[];
```
```suggestion
    utxos: TransactionUnspentOutput[];
```

mkazlauskas replied:
The [spec](https://cips.cardano.org/cips/cip2/#results) says "A set of **inputs**, equivalent to a subset of the initial UTxO set.". Please confirm this needs to be renamed.

rhyslbw replied:
Ok, let's leave it as-is

rhyslbw commented on packages/cip2/src/types.ts:
```diff
@@ -0,0 +1,73 @@
+  utxo: TransactionUnspentOutput[],
+  outputs: TransactionOutput[],
+  change: Value[]
+) => bigint;
```
```suggestion
) => Promise<number>;
```

rhyslbw replied:
We need to fully construct the transaction in order to estimate the fee, which requires a call to: https://github.com/input-output-hk/cardano-js-sdk/blob/e37fce63b00645897ae1ef6e0220c75ec133d163/packages/wallet/src/KeyManagement/KeyManager.ts#L12-L14

mkazlauskas replied:
That's unfortunate, it makes the entire `InputSelector.select` async

rhyslbw replied:
We could also just change the interface, as the in-memory implementation isn't async:

https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/in-memory-key-manager/src/InMemoryKey.ts#L43-L50

I'll double-check to see what we need to support hardware wallets

rhyslbw replied:
Unfortunately it's an async operation:

https://github.com/cardano-foundation/ledgerjs-hw-app-cardano/blob/47927d42e9b2b0a70463a719cbd745a84cbfce81/src/interactions/signTx.ts#L41-L56

rhyslbw commented on packages/cip2/test/util.ts:
```diff
@@ -0,0 +1,209 @@
+// Review: these are utilities used by tests.
```
Let's leave it here for now, and consider introducing a `util-dev` package

rhyslbw commented on packages/cip2/test/util.ts:
```diff
@@ -0,0 +1,209 @@
+  const getTotalInputAmounts = (results: SelectionResult): ValueQuantities =>
+    results.selection.inputs
+      .map((input) => input.output().amount())
+      // eslint-disable-next-line unicorn/no-array-reduce
```
Also can disable this rule

mkazlauskas commented on packages/cip2/test/jest.setup.js:
```diff
@@ -0,0 +1,4 @@
+// TODO: jest environment is not happy with 'lodash-es' exports.
```
Can this TODO be merged as-is?

rhyslbw replied:
Yes

mkazlauskas commented:
@rhyslbw I think I addressed all the comments, let me know if I missed something. Please take a look at [this](https://github.com/input-output-hk/cardano-js-sdk/pull/84/files#diff-5f90861d94ca998bd5eaf8b0fcc4d8f89446ae8fc72da57e787ec56fe7361292R58) too.

I didn't move the Ogmios<-->CSL value conversion util functions out of cip2 yet, because we need CSL to update lovelace type to bigint first. Left a comment about that.

Also I think we should move OgmiosToCardanoWasm to core package and remove core package dependency from it. Core package has a dependency on CSL package.

rhyslbw commented:
> Also I think we should move OgmiosToCardanoWasm to core package and remove core package dependency from it. Core package has a dependency on CSL package.

Done in d13513fff3e1ec91cbd4631f348b1bbcf5d0e3a8, which fixes the CI failure

> Please take a look at this too.

I've asked Jonathan for advice. Cardano does have a transaction size limit, so it's safe to assume we need it.

rhyslbw commented on packages/core/src/Asset/util.ts:
```diff
@@ -1,33 +1,24 @@
+export const createAssetId = (scriptHash: ScriptHash, assetName: AssetName): string =>
+  Buffer.from(scriptHash.to_bytes()).toString('hex') + textDecoder.decode(assetName.name());
+
+export const parseAssetId = (id: string, csl: CardanoSerializationLib) => {
```
Let's always us `assetId` to be explicit

rhyslbw commented on packages/cip2/src/util.ts:
```diff
@@ -18,61 +18,52 @@ export const transactionOutputsToArray = (outputs: CSL.TransactionOutputs): CSL.
+  const assetIds = Object.keys(assets);
+  if (assetIds.length > 0) {
+    const multiasset = csl.MultiAsset.new();
+    for (const id of assetIds) {
```
`id` -> `assetId`

rhyslbw commented on packages/core/src/Asset/util.ts:
```diff
@@ -1,33 +1,24 @@
+ * @returns {string} concatenated policy id and asset name
+ */
+export const createAssetId = (scriptHash: ScriptHash, assetName: AssetName): string =>
+  Buffer.from(scriptHash.to_bytes()).toString('hex') + textDecoder.decode(assetName.name());
```
`assetName` shouldn't be decoded. The only time we would decode is for presentation purposes, but there's no enforcement in the ledger so it's not guaranteed to be UTF-8.

mkazlauskas replied:
I see. We still need to stringify it. How about `[...assetName].join('')`?

rhyslbw replied:
```
Buffer.from(asset.name()).toString("hex");
```

rhyslbw commented on packages/cip2/src/util.ts:
```diff
@@ -0,0 +1,69 @@
+// This is equivalent to OgmiosToCardanoWasm.value(), so they should be merged.
+// Note that current implementation of OgmiosToCardanoWasm
+// imports nodejs version of CSL, so it cannot be used in browser.
+// Also note that this function needs to know what bech32 prefix was used in order to recreate
```
We can hard-code the prefix as `asset` as it's the same across all networks

https://cips.cardano.org/cips/cip5/

rhyslbw reviewed (approved):
Excellent work @mkazlauskas :tada:
